### PR TITLE
Implements petautobonus script commands

### LINKED ISSUE #7038: Autobonus on Pet_db.yml Effect don't work
* **rAthena Hash**: Latest

* **Client Date**: 2021-11-17

* **Server Mode**: Renewal

* **Description of Issue**: autobonus on pet_db.yml effect don't work
  * Result: autobonus on pet bonus script didn't work.
  * Expected Result: autobonus on pet bonus script should work.
  * How to Reproduce: try to use pet MARIONETTE (Marionette_Egg ID : 9043) then change
  * `autobonus "{ bonus2 bSubEle,Ele_Neutral,2; heal 100,0; }",100,5000,BF_SHORT|BF_NORMAL;`
  * to
  * `autobonus "{ bonus bDex,1000; }",1000,5000,BF_SHORT|BF_NORMAL;`
  * type command `@petfriendly 1000` and yes change **pet.conf** `pet_status_support: yes`
  * Official Information: 
  

* **Modifications that may affect results**: try to attack dummy but effect autobonus don't work

pet png 64daa6f1049d5bb9032de36534345843

### PR DESCRIPTION
**Pull request by aleos89**
* **Addressed Issue(s)**: Fixes #7038

* **Server Mode**: Pre-renewal and Renewal

* **Description of Pull Request**: 
  * Implements script commands: petautobonus, petautobonus2, and petautobonus3.
  * This allows pets to utilize the same autobonus features that players can use without having to check for the equipped item position.

### CODE REVIEW

**Review by null6385 — APPROVED** [review]

File: src/map/pet.cpp [hunk 1]
```diff
@@ -2332,6 +2333,98 @@ void pet_evolution(struct map_session_data *sd, int16 pet_id) {
...
+ * Remove petautobonus timer from player.
+ */
+TIMER_FUNC(pet_endautobonus) {
+	map_session_data *sd = map_id2sd(id);
+	std::shared_ptr<s_petautobonus> *bonus = (std::shared_ptr<s_petautobonus> *)data;
+
+	nullpo_ret(sd);
+	nullpo_ret(bonus);
+
+	if (bonus->get()->timer == tid)
```
Comment (Lemongrass3110) [inline src/map/pet.cpp:2421]: Is the get() in these two statements really necessary?

File: src/map/script.cpp [hunk 2]
```diff
@@ -9957,6 +9978,97 @@ BUILDIN_FUNC(autobonus3)
...
+		ShowError("buildin_%s: Requires an active pet.\n", command);
+		return SCRIPT_CMD_FAILURE; // No pet attached to player
+	}
+
+	std::string bonus_script = script_getstr(st, 2);
+	int16 rate = script_getnum(st, 3);
+	uint32 dur = script_getnum(st, 4);
+
+	if (!rate || !dur || bonus_script.empty())
+		return SCRIPT_CMD_SUCCESS;
```
Comment (Lemongrass3110) [inline src/map/script.cpp:9999]:
```suggestion
		return SCRIPT_CMD_FAILURE;
```

File: src/map/script.cpp [hunk 3]
```diff
@@ -9957,6 +9978,97 @@ BUILDIN_FUNC(autobonus3)
...
+		ShowError("buildin_petautobonus3: Requires an active pet.\n");
+		return SCRIPT_CMD_FAILURE; // No pet attached to player
+	}
+
+	std::string bonus_script = script_getstr(st, 2);
+	int16 rate = script_getnum(st, 3);
+	uint32 dur = script_getnum(st, 4);
+
+	if (!rate || !dur || bonus_script.empty())
+		return SCRIPT_CMD_SUCCESS;
```
Comment (Lemongrass3110) [inline src/map/script.cpp:10038]:
```suggestion
		return SCRIPT_CMD_FAILURE;
```

File: src/map/skill.cpp [hunk 4]
```diff
@@ -2687,6 +2705,15 @@ int skill_counter_additional_effect (struct block_list* src, struct block_list *
 		}
 	}
 
+	// Pet Autobonus when attacked
+	if (dstsd != nullptr && dstsd->pd != nullptr && dstsd->pd->autobonus2 != nullptr && !status_isdead(bl) && !(skill_id && skill_get_nk(skill_id, NK_NODAMAGE))) {
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:2709]: Would be better to merge the two ifs for normal autobonus and these.
Like combine one outer if with all checks that apply to both and then have inner ifs for pc bonus and pet bonus.
This comment most likely to all of these checks.

File: src/map/pet.hpp [hunk 5]
```diff
@@ -165,6 +201,7 @@ struct pet_data {
 	struct pet_skill_attack* a_skill;
 	struct pet_skill_support* s_skill;
 	struct pet_loot* loot;
+	std::shared_ptr<s_petautobonus> autobonus, autobonus2, autobonus3;
```
Comment (Lemongrass3110) [inline src/map/pet.hpp:204]: Is there a reason why we only allow one autobonus for pets and a whole vector for players?

**Review by EditorFc — APPROVED** [review]

**Review by Lemongrass3110 — APPROVED** [review]

File: src/map/pet.cpp [hunk 6]
```diff
@@ -2332,6 +2333,114 @@ void pet_evolution(struct map_session_data *sd, int16 pet_id) {
...
+
+	std::shared_ptr<s_petautobonus> entry = std::make_shared<s_petautobonus>();
+
+	entry->rate = cap_value(rate, -10000, 10000);
+	entry->duration = dur;
+	entry->timer = INVALID_TIMER;
+	entry->atk_type = flag;
+	entry->bonus_script = script;
+	if (!other_script.empty())
+		entry->other_script = other_script;
```
Comment (aleos89) [inline src/map/pet.cpp:2378]:
```suggestion
	entry->other_script = other_script;
```

**Pull request merged**